### PR TITLE
init.h: restore designated initializers in SYS_INIT_NAMED() v2

### PR DESCRIPTION
**Pull request by marc-hb**
As seen in the ~PR #68125~ and #69411 discussions, commit 19a33c788413 ("init: adjust the SYS_INIT dev field init to play nice with older compilers") entirely threw away designated initializers in SYS_INIT_NAMED() to avoid compatibility issues across toolchains.

One key aspect that was probably missed at the time: C and C++ are two different languages and this is especially true with respect to designated initializers.

Designated initializers provide safer and more readable code, especially in their much stricter C++ version. So use an #ifdef to restore them in SYS_INIT_NAMED() thanks to a small braces difference between C and C++.

### CODE REVIEW

**Comment by marc-hb** [comment]
Alternative to #69411 recommended by @keith-packard . Only the `#ifdef` is different compared to #69411.

  ------

Some Github workflows are timing out:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/8054336494/job/21998974926?pr=69486

This is not specific to this PR, for instance this pure doxygen fix experienced a timeout too:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/8053793552/job/21998901961

https://github.com/zephyrproject-rtos/zephyr/actions/workflows/twister.yaml

**Review by keith-packard — DISMISSED** [review]
Yeah, this should work great. Thanks for the discussion; I learned quite a bit!

**Review by cfriedt — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@gmarull ping

File: include/zephyr/init.h [hunk 1]
```diff
@@ -151,6 +151,39 @@ struct init_entry {
...
+#else
+
+/* When using -std=c++20 or higher, g++ (v12.2.0) reject braces for
+ * initializing anonymous unions because it is technically a mix of
+ * designated and not designated initializers which is not allowed in
+ * C++. Interestingly, the _same_ g++ version does accept the braces above
+ * when using -std=c++17 or lower!
+ * This Zephyr test compares different C++ -std=... values:
+ *
+ *    ./scripts/twister -v -b -p qemu_cortex_a53 -T tests/lib/cpp/
```
Comment (gmarull) [inline include/zephyr/init.h:181]: please drop CLI snippets from here, maybe just reference the test.

**Review by keith-packard — APPROVED** [review]

File: include/zephyr/init.h (hunk 1 above)
Comment (marc-hb) [reply include/zephyr/init.h]: > please drop CLI snippets from here, maybe just reference the test.

Done!

Small, pure comment change; @cfriedt can you please re-approve?

**Review by cfriedt — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Review by kv2019i — APPROVED** [review]

**Pull request merged**